### PR TITLE
Interface2

### PR DESCRIPTION
**Pull request by mloubout**
Reopening interface enhancement, WIP but mostly there

- '(u + v.dx).dy' and things like that possible
- staggered no is defined as `staggered = x` for example, not more `(0,1,0)` and `None` corresponds to cartesian grids. Need to add check to make sure all fields are either staggered or non-staggered
- FD shortcuts are no only created if called to avoid hundreds if not thousands of calls to `setattr` during FD manipulation

- Some misc

### CODE REVIEW

File: examples/seismic/elastic/operators.py [hunk 1]
```diff
@@ -26,56 +87,111 @@ def ForwardOperator(model, source, receiver, space_order=4,
...
+    src_expr = src_xx + src_zz
+    if model.grid.dim == 3:
+        src_yy = src.inject(field=tyy.forward, expr=src * s, offset=model.nbpml)
+        src_expr += src_yy
 
     # Create interpolation expression for receivers
-    rec_term1 = rec1.interpolate(expr=txx, offset=model.nbpml)
-    rec_term2 = rec2.interpolate(expr=tzz, offset=model.nbpml)
+    rec_term1 = rec1.interpolate(expr=tzz, offset=model.nbpml)
+    if model.grid.dim == 2:
```
Comment (FabioLuporini) [inline examples/seismic/elastic/operators.py]: this makes me think that we could to introduce the concept of ``VectorFunction``, so that we just carry ``v`` around. Then automatically ``v = [vx]``  in 1D, ``v = [vx, vy]`` in 2D, etc., which means that ``rec_expr`` (and many, many other operations typical of problem setup!) don't need to be specialised depending on whether it's 1/2/3D... it'd just be ``rec_expr = v.sum``

Of course this is just food for thought, no need to do anything in this PR which is already quite big

File: devito/finite_differences/differentiable.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+    def _merge_fd_properties(self, other):
+        merged = {}
+        merged["space_order"] = np.min([getattr(self, 'space_order', 100) or 100,
+                                        getattr(other, 'space_order', 100)])
+        merged["time_order"] = np.min([getattr(self, 'time_order', 100) or 100,
+                                       getattr(other, 'time_order', 100)])
+        merged["indices"] = tuple(set(self.indices + getattr(other, 'indices', ())))
+        merged["fd"] = dict(getattr(self, 'fd', {}), **getattr(other, 'fd', {}))
+        merged["staggered"] = self.staggered
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: should we not check that they are actually compatible? 
that's why I was wondering about the use of the word "merge" .

File: devito/finite_differences/differentiable.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        return self._grid
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    def __getattr__(self, name):
+        if name == 'fd' or name == '_fd':
+            raise AttributeError()
+        if name in self.fd.keys():
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: reminder: no need for keys()

File: devito/grid.py [hunk 4]
```diff
@@ -201,3 +201,28 @@ def __setstate__(self, state):
...
+
+def staggered(function):
+    """
+    For a given function and it's staggered property `staggered`
+    returns a tuple of 0 or 1 for each dimension:
+    0 if not staggered in the dimension
+    1 if staggered in the dimension
+    """
+    s = (lambda x: x if type(x) is tuple else (x,))(function.staggered)
+    if function.staggered in ['node', None]:
```
Comment (FabioLuporini) [inline devito/grid.py]: 'node' and 'cell' should be defined through ``Tag``. ``git grep Tag`` for some examples.

File: devito/function.py [hunk 5]
```diff
@@ -538,7 +532,7 @@ def _arg_check(self, args, intervals):
         ['dtype', 'grid', 'staggered', 'initializer']
 
 
-class Function(TensorFunction):
+class Function(TensorFunction, Differentiable):
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/function.py]: doesn't call the `Differential` `super` so not sure

File: devito/function.py [hunk 6]
```diff
@@ -596,18 +590,20 @@ class Function(TensorFunction):
 
     .. note::
 
-       The tuple :param staggered: contains a ``1`` in each dimension
-       entry that should be staggered, and ``0`` otherwise. For example,
-       ``staggered=(1, 0, 0)`` entails discretization on horizontal edges,
-       ``staggered=(0, 0, 1)`` entails discretization on vertical edges,
-       ``staggered=(0, 1, 1)`` entails discretization side facets and
-       ``staggered=(1, 1, 1)`` entails discretization on cells.
+       The dimension or String :param staggered: defines the dimension in which the
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/function.py]: `Dimension`

File: devito/function.py [hunk 7]
```diff
@@ -596,18 +590,20 @@ class Function(TensorFunction):
...
+       ``staggered=(x, y)`` entails discretization xy facets and
+       ``staggered='node'`` entails discretization on node
+       ``staggerd='cell'`` entails discretization on cell.
     """
 
     is_Function = True
 
     def __init__(self, *args, **kwargs):
         if not self._cached():
+            self._kwargs = kwargs
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/function.py]: gone, leftover

File: devito/operator.py [hunk 8]
```diff
@@ -71,7 +72,7 @@ def __init__(self, expressions, **kwargs):
         self._func_table = OrderedDict()
 
         # Expression lowering: indexification, substitution rules, specialization
-        expressions = [indexify(i) for i in expressions]
+        expressions = [indexify(to_expr(i)) for i in expressions]
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/operator.py]: No need for `Differential` anymore in the compiler, so get sympy expr only at this point

File: devito/symbolics/extended_sympy.py [hunk 9]
```diff
@@ -109,6 +109,29 @@ def __str__(self):
     __repr__ = __str__
 
 
+class ExprDiv(sympy.Expr):
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/symbolics/extended_sympy.py]: `IntDiv` uses `Integer` to enforce integer indices for subsampling, `ExprDiv` is general expression and `Integer(Expr)` does not exist

File: examples/seismic/elastic/operators.py (hunk 1 above)
Comment (mloubout) [reply examples/seismic/elastic/operators.py]: >rec_expr = v.sum

Not thecase in 3D, only `vx+vz` `vy` is ignored. 

>`VectorFunction`

That would be great but requires (estimates :D) same size PR by itself

File: examples/seismic/elastic/wavesolver.py [hunk 10]
```diff
@@ -72,28 +69,25 @@ def forward(self, src=None, rec1=None, rec2=None, vp=None, vs=None, rho=None,
...
+            kwargs['txy'] = txy
+            kwargs['tyz'] = tyz
         # Pick m from model unless explicitly provided
         vp = vp or self.model.vp
         vs = vs or self.model.vs
         rho = rho or self.model.rho
         # Execute operator and return wavefield and receiver data
-        summary = self.op_fwd(save).apply(src=src, rec1=rec1, vx=vx, vz=vz, txx=txx,
-                                          tzz=tzz, txz=txz, vp=vp, vs=vs, rho=rho,
+        summary = self.op_fwd(save).apply(src=src, rec1=rec1, vp=vp, vs=vs, rho=rho,
```
[the remark this answers is not in the document]
Comment (mloubout) [reply examples/seismic/elastic/wavesolver.py]: `if 2d/3d` for the old way but can change

File: examples/seismic/utils.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
+import numpy as np
+
+
+__all__ = ['smooth10']
+
+
+# Velocity models
+def smooth10(vel, shape):
```
[the remark this answers is not in the document]
Comment (mloubout) [reply examples/seismic/utils.py]: yep

File: devito/finite_differences/differentiable.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+
+    def __init__(self, expr, **kwargs):
+        # Recover the list of possible FD shortcuts
+        self._dtype = kwargs.get('dtype')
+        self._space_order = kwargs.get('space_order')
+        self._time_order = kwargs.get('time_order', 0)
+        self._indices = kwargs.get('indices', ())
+        self._staggered = kwargs.get('staggered')
+        self._grid = kwargs.get('grid')
+        # Generate FD shortcuts for expression or copy from input
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: obsolete comment?

File: devito/finite_differences/differentiable.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+            assert len(args) == 1
+            expr = args[0]
+            new_obj = sympy.Expr.__new__(cls, expr)
+            # Initialization
+            new_obj.__init__(expr, **kwargs)
+            return new_obj
+        return sympy.Expr.__new__(cls, *args)
+
+    def __init__(self, expr, **kwargs):
+        # Recover the list of possible FD shortcuts
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: obsolete comment?

File: devito/finite_differences/differentiable.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+        self._dtype = kwargs.get('dtype')
+        self._space_order = kwargs.get('space_order')
+        self._time_order = kwargs.get('time_order', 0)
+        self._indices = kwargs.get('indices', ())
+        self._staggered = kwargs.get('staggered')
+        self._grid = kwargs.get('grid')
+        # Generate FD shortcuts for expression or copy from input
+        self._fd = kwargs.get('fd', {})
+        # Save kwargs
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: obsolete comment?

File: devito/finite_differences/differentiable.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+        self._expr = expr
+
+    def __getattr__(self, name):
+        if name == 'fd':
+            raise AttributeError()
+        if name in self.fd.keys():
+            return self.fd[name][0](self)
+        return self.__getattribute__(name)
+
+    def xreplace(self, rule):
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py]: then get _expr and call the superclass' xreplace via super?

File: devito/finite_differences/differentiable.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+        merged["staggered"] = self.staggered
+        merged["dtype"] = self.dtype
+        return merged
+
+    def __add__(self, other):
+        return Differentiable(sympy.Add(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    __iadd__ = __add__
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: I'm not sure __iadd__ and __radd__ should be the same as __add__. Is it what happens in sympy typically? __iadd__ is += 

same for sub, mul etc

File: devito/symbolics/extended_sympy.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/symbolics/extended_sympy.py]: ah, right, I was confused by the docstring which says "a support type for integer division..."

File: examples/seismic/elastic/wavesolver.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply examples/seismic/elastic/wavesolver.py]: can you not just provide None for the unused Functions? does the Operator complain?

File: devito/finite_differences/differentiable.py [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    sum of functions, product of function or FD approximation and
+    provides FD shortcuts for such expressions
+    """
+    _op_priority = 100.0
+
+    def __new__(cls, *args, **kwargs):
+        if cls == Differentiable:
+            assert len(args) == 1
+            expr = args[0]
+            new_obj = sympy.Expr.__new__(cls, expr)
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: out of curiosity: what happens if you rather do:
```sympy.Expr.__new__(sympy.Expr, expr)```
and then
```new_obj._space_order = ...
new_obj._dtype = ...
...
return new_obj
```
do we get "true" SymPy objects but with additional attributes? or do you actually get exceptions when trying to add new fields?

Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py]: most likely an exception . Scratch this.

File: devito/finite_differences/differentiable.py (hunk 16 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py]: Sympy expressions are immutable, and operation on it should return a new one.

File: examples/seismic/elastic/wavesolver.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (mloubout) [reply examples/seismic/elastic/wavesolver.py]: Yes it does, `Operator` complains for any kwargs not in `arguments` no matter the value assigned to it.

File: devito/finite_differences/differentiable.py (hunk 16 above)
Comment (FabioLuporini) [reply devito/finite_differences/differentiable.py]: it appears that += actually has an effect:
```
In [1]: from sympy import *
...
In [5]: a=Symbol('a')
In [6]: b=Symbol('b')
In [7]: a + b
Out[7]: a + b
In [8]: a += b
In [9]: a
Out[9]: a + b
```
note that the content of ``a`` has actually changed

as for __radd__, __rmul__, __rdiv__ ... we should swap the operands. See for example here: https://github.com/sympy/sympy/blob/master/sympy/core/expr.py#L124
not sure, however, what are the practical implications of this in SymPy

File: devito/finite_differences/differentiable.py [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+
+    def __rtruediv__(self, other):
+        return Differentiable(sympy.Mul(*[getattr(other, '_expr', other),
+                                          getattr(self, '_expr', self)**(-1)]),
+                              **self._merge_fd_properties(other))
+
+    __floordiv__ = __truediv__
+    __rdiv__ = __rtruediv__
+    __div__ = __truediv__
+    __rfloordiv__ = __rtruediv__
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: Note sure where to leave this comment... anyway...
we're missing __mod__, __int__, and __float__ at least

File: devito/finite_differences/differentiable.py [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+    provides FD shortcuts for such expressions
+    """
+    # Set the operator priority higher than Sympy (10.0) to force the overloaded
+    # operators to be used
+    _op_priority = 100.0
+
+    def __new__(cls, *args, **kwargs):
+        if cls == Differentiable:
+            assert len(args) == 1
+            expr = args[0]
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: I have an idea, I think. You can check here whether:
```
if expr.is_Function
```
in which case, you just
```
    return expr
```
this should avoid the proliferation of ``if self.is_Function`` in the various methods

File: devito/grid.py [hunk 20]
```diff
@@ -201,3 +201,28 @@ def __setstate__(self, state):
         for k, v in state.items():
             setattr(self, k, v)
         self._distributor = Distributor(self.shape, self.dimensions)
+
+
+def staggered(function):
```
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/grid.py]: I suggest we move this function into ``TensorFunction``, and turn it into a (overridable, should we need to do something different for SparseFunctions...) ``__staggered_setup__`` method.

File: devito/grid.py [hunk 21]
```diff
@@ -201,3 +201,28 @@ def __setstate__(self, state):
...
+    returns a tuple of 0 or 1 for each dimension:
+    0 if not staggered in the dimension
+    1 if staggered in the dimension
+    """
+    s = (lambda x: x if type(x) is tuple else (x,))(function.staggered)
+    if function.staggered in ['node', None]:
+        return tuple(0 for _ in function.indices)
+    if function.staggered == 'cell':
+        return tuple(1 for _ in function.indices)
+    staggered = ()
```
Comment (FabioLuporini) [inline devito/grid.py]: BTW, this should be a list...

File: devito/grid.py [hunk 22]
```diff
@@ -201,3 +201,28 @@ def __setstate__(self, state):
...
+    s_neg = tuple([-ss for ss in s])
+    for d in function.indices:
+        if d in s:
+            staggered += (1,)
+        elif d in s_neg:
+            staggered += (-1,)
+        else:
+            staggered += (0,)
+
+    return staggered
```
Comment (FabioLuporini) [inline devito/grid.py]: ...and casted to tuple here

File: devito/operator.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/operator.py]: would things break without this ``to_expr`` ?

File: devito/symbolics/extended_sympy.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/symbolics/extended_sympy.py]: meaning that we should update the docstring

File: devito/finite_differences/differentiable.py [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+            return sympy.Number(1)
+
+    def __neg__(self):
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self), -1]),
+                              **self._merge_fd_properties(None))
+
+    def __str__(self):
+        if self.is_Function:
+            return super(Differentiable, self).__str__()
+        return to_expr(self._expr).__str__()
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: perhaps we don't need to wrap it with ``to_expr``

File: devito/types.py [hunk 24]
```diff
@@ -606,7 +606,9 @@ def indexify(self, indices=None):
         if indices is not None:
             return Indexed(self.indexed, *indices)
 
-        subs = dict([(i.spacing, 1) for i in self.indices])
+        # Only replace spacing -> 1 if used as index
```
Comment (FabioLuporini) [inline devito/types.py]: why this change?

File: devito/finite_differences/differentiable.py (hunk 16 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py]: Hum odd, I can rewrite `iadd,imul,..` to just change `_expr` in place and keep same object, fine with me.

for the `r` ones, additions are commutative so doesn't really matter but can change it.

File: devito/finite_differences/differentiable.py (hunk 23 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py]: probably, was getting some `Differentiable` in the str at some point so had to put it.

File: devito/grid.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/grid.py]: Will think if not too intricated.

File: devito/operator.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/operator.py]: Yes, was doing werid stuff without it

File: devito/types.py (hunk 24 above)
Comment (mloubout) [reply devito/types.py]: I "fixed" `SparseFunction` tthat was forcing `indexify` in inject/interpolate to make this function happy.

File: devito/function.py [hunk 25]
```diff
@@ -46,7 +43,7 @@ class Constant(AbstractCachedSymbol):
 
     def __init__(self, *args, **kwargs):
         if not self._cached():
-            self.dtype = kwargs.get('dtype', np.float32)
+            self._dtype = kwargs.get('dtype', np.float32)
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/function.py]: `AbstractCachedSymbol` doesn't have a dtype

[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/function.py]: But types.Symbol and types.Scalar do. We should then factor this into ``AbstractCachedSymbol``

[the remark this answers is not in the document]
Comment (mloubout) [reply devito/function.py]: We can't, `Scalar` defaults to `float` while `Symbol` defaults to `int`, would be a huge diggin to track everywhere it's used to make it work and this PR already big enough.

[the remark this answers is not in the document]
Comment (FabioLuporini) [reply devito/function.py]: no worries, I'm touching those classes in another PR, I'll find a way later on

File: devito/finite_differences/differentiable.py [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+        return Differentiable(sympy.Mul(*[getattr(self, '_expr', self),
+                                          getattr(other, '_expr', other)]),
+                              **self._merge_fd_properties(other))
+
+    def __imul__(self, other):
+        self._expr = sympy.Mul(*[getattr(self, '_expr', self),
+                                 getattr(other, '_expr', other)])
+        return self
+
+    __rmul__ = __mul__
```
Comment (FabioLuporini) [inline devito/finite_differences/differentiable.py]: (maybe nitpicking, maybe not): floating point operations are not commutative

File: devito/types.py [hunk 27]
```diff
@@ -369,6 +370,11 @@ def __init__(self, *args, **kwargs):
 
     __hash__ = Cached.__hash__
 
+    @classmethod
+    def __staggered_setup__(cls, obj, **kwargs):
+        """Extract the object indices from ``kwargs``."""
```
Comment (FabioLuporini) [inline devito/types.py]: update pasted docstring

File: devito/function.py [hunk 28]
```diff
@@ -193,6 +187,40 @@ def __dtype_setup__(cls, **kwargs):
...
+                staggered += [1]
+            elif d in s_neg:
+                staggered += [-1]
+            else:
+                staggered += [0]
+
+        return tuple(staggered)
+
+    @property
+    def staggered(self):
```
Comment (FabioLuporini) [inline devito/function.py]: this should go in ``AbstractCachedFunction``

File: devito/function.py [hunk 29]
```diff
@@ -193,6 +187,40 @@ def __dtype_setup__(cls, **kwargs):
...
+    @classmethod
+    def __staggered_setup__(cls, func, **kwargs):
+        """
+        For a given function and it's staggered property `staggered`
+        returns a tuple of 0 or 1 for each dimension:
+        0 if not staggered in the dimension
+        1 if staggered in the dimension
+        """
+        stagger = kwargs.get('staggered')
+        s = (lambda x: x if type(x) is tuple else (x,))(stagger)
```
Comment (FabioLuporini) [inline devito/function.py]: perhaps ``as_tuple`` helps here?

File: devito/function.py [hunk 30]
```diff
@@ -193,6 +187,40 @@ def __dtype_setup__(cls, **kwargs):
...
+        if stagger is None:
+            func.is_Staggered = False
+            return tuple(0 for _ in func.indices)
+        func.is_Staggered = True
+        if stagger == 'node':
```
Comment (FabioLuporini) [inline devito/function.py]: is this just a special case of the loop below?

File: devito/function.py [hunk 31]
```diff
@@ -193,6 +187,40 @@ def __dtype_setup__(cls, **kwargs):
...
+            return tuple(0 for _ in func.indices)
+        if stagger == 'cell':
```
Comment (FabioLuporini) [inline devito/function.py]: is this just a special case of the loop below?

File: devito/types.py [hunk 32]
```diff
@@ -348,6 +348,7 @@ def __new__(cls, *args, **kwargs):
             newobj._indices = indices
             newobj._shape = cls.__shape_setup__(**kwargs)
             newobj._dtype = cls.__dtype_setup__(**kwargs)
+            newobj._staggered = cls.__staggered_setup__(newobj, **kwargs)
```
Comment (FabioLuporini) [inline devito/types.py]: don't pass ``newobj``, just have ``__staggered_setup__`` return what you want to store in ``_staggered``

File: devito/types.py [hunk 33]
```diff
@@ -369,6 +370,11 @@ def __init__(self, *args, **kwargs):
 
     __hash__ = Cached.__hash__
 
+    @classmethod
+    def __staggered_setup__(cls, obj, **kwargs):
```
Comment (FabioLuporini) [inline devito/types.py]: from the previous comment, no need for ``obj`` here

File: devito/function.py [hunk 34]
```diff
@@ -193,6 +187,40 @@ def __dtype_setup__(cls, **kwargs):
         else:
             return np.float32
 
+    @classmethod
+    def __staggered_setup__(cls, func, **kwargs):
```
Comment (FabioLuporini) [inline devito/function.py]: no need for ``func`` here. See comments below in ``types.py``.

File: devito/function.py [hunk 35]
```diff
@@ -193,6 +187,40 @@ def __dtype_setup__(cls, **kwargs):
...
+        returns a tuple of 0 or 1 for each dimension:
+        0 if not staggered in the dimension
+        1 if staggered in the dimension
+        """
+        stagger = kwargs.get('staggered')
+        s = (lambda x: x if type(x) is tuple else (x,))(stagger)
+        if stagger is None:
+            func.is_Staggered = False
+            return tuple(0 for _ in func.indices)
+        func.is_Staggered = True
```
Comment (FabioLuporini) [inline devito/function.py]: no need to set anything here. ``is_Staggered`` is better off as a property such as ``return any(i == 1 for i in self.staggered``)

File: devito/finite_differences/differentiable.py (hunk 26 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py]: It's on symbolic  expressions  so should be

File: devito/types.py (hunk 32 above)
Comment (mloubout) [reply devito/types.py]: Necessary, needs to access `obj.indices` in the function

Comment (FabioLuporini) [reply devito/types.py]: ah, so don't make it a classmethod. For example, look at __halo_setup__

File: tests/test_ir.py [hunk 36]
```diff
@@ -524,7 +524,7 @@ def test_iteration_parallelism_2d(self, exprs, atomic, parallel):
         for i, e in enumerate(list(exprs)):
             exprs[i] = eval(e)
 
-        op = Operator(exprs, dle='advanced')
+        op = Operator(exprs, dle='openmp')
```
Comment (FabioLuporini) [inline tests/test_ir.py]: wait, what?

File: devito/equation.py [hunk 37]
```diff
@@ -89,4 +91,7 @@ def solve(eq, target, **kwargs):
     # turnaround time
     kwargs['rational'] = False  # Avoid float indices
     kwargs['simplify'] = False  # Do not attempt premature optimisation
+    if hasattr(eq, 'lhs'):
```
Comment (FabioLuporini) [inline devito/equation.py]: what? recently I've seen this ``hasattr`` too often.

can you not just ``if eq.is_Equality`` ? and btw, what else should/could it be?

File: tests/test_ir.py (hunk 36 above)
Comment (mloubout) [reply tests/test_ir.py]: with `advanced` cache blocking dims appear which make it fail, 

># Use 'openmp' here instead of 'advanced' to disable loop blocking
line 584 for 3D, should be same here

File: devito/equation.py (hunk 37 above)
Comment (mloubout) [reply devito/equation.py]: >and btw, what else should/could it be?
it can be a single expression as solve defaults to `Eq(expr, 0)` if called as `solve(expr)` in which case `expr` does not have `is_Equality` attribute

File: tests/test_ir.py (hunk 36 above)
Comment (FabioLuporini) [reply tests/test_ir.py]: then it's a bug to be fixed

File: devito/equation.py (hunk 37 above)
Comment (FabioLuporini) [reply devito/equation.py]: why not? all sympy expressions, even symbols, have ``is_Equality``

File: tests/test_ir.py (hunk 36 above)
Comment (mloubout) [reply tests/test_ir.py]: It;s not a bug, that's what's needed for that test

Comment (FabioLuporini) [reply tests/test_ir.py]: ahhhhhh sorry now I see what you mean!

File: devito/finite_differences/differentiable.py (hunk 19 above)
Comment (mloubout) [reply devito/finite_differences/differentiable.py]: To avoid duplicates of `add,mul,...` makes evrything lots easier and less messy to have `Function` being differentiable

File: devito/finite_differences/finite_difference.py [hunk 38]
```diff
@@ -0,0 +1,401 @@
...
+                            fd_order=dim_order, stagger=centered)
+            name_fd = 'd%sc' % name
+            desciption = 'centered derivative staggered w.r.t dimension %s' % d
+
+            derivatives[name_fd] = (deriv, desciption)
+
+    return derivatives
+
+
+def initialize_derivatives(function):
```
Comment (FabioLuporini) [inline devito/finite_differences/finite_difference.py]: Reminder: we should call generate_fd_shortcuts from within Function.__init__

File: devito/grid.py [hunk 39]
```diff
@@ -49,21 +49,20 @@ class Grid(ArgProvider):
 
        .. code-block:: python
 
-          x ^
-            |
-            |           origin + extent
+                      x
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/grid.py]: Nope still x horizontal and y vertical, only look weird when plotting because of row major python

File: devito/equation.py [hunk 40]
```diff
@@ -27,6 +27,9 @@ def xreplace(self, rules):
         return self.func(self.lhs.xreplace(rules), self.rhs.xreplace(rules),
                          region=self._region)
 
+    def __str__(self):
```
[the remark this answers is not in the document]
Comment (mloubout) [reply devito/equation.py]: Leftover from Differentiable not printing nicelly inside Eq, should be dropable now

File: devito/finite_differences/finite_difference.py [hunk 41]
```diff
@@ -0,0 +1,401 @@
...
+            deriv = partial(first_derivative, order=space_fd_order,
+                            dim=d, side=left)
+            name_fd = 'd%sl' % name
+            desciption = 'left first order derivative w.r.t dimension %s' % d
+            derivatives[name_fd] = (deriv, desciption)
+            # right
+            deriv = partial(first_derivative, order=space_fd_order,
+                            dim=d, side=right)
+            name_fd = 'd%sr' % name
+            desciption = 'left first order derivative w.r.t dimension %s' % d
```
Comment (mloubout) [inline devito/finite_differences/finite_difference.py]: I thought I fixed that, commits mix up must have brought it back
